### PR TITLE
Transform

### PR DESCRIPTION
**Pull request by ntut-rick**

### CODE REVIEW

File: include/Util/Transform.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    Transform(glm::mat3 mat3) : m_mat3(mat3) {};
+
+    Transform Rotate(float degrees) const { return RotateByRadians(glm::radians(degrees)); }
+    Transform RotateByRadians(float radians) const;
+
+    Transform Translate(const glm::vec2& translation) const;
+    Transform Translate(float x, float y) const { return Translate(glm::vec2(x,y)); }
+
+    Transform Scale(const glm::vec2& scale) const;
+    Transform Scale(float x, float y) const { return Scale(glm::vec2(x,y)); }
```
Comment (NOOBDY) [inline include/Util/Transform.hpp:20]: I don't think overloading here is necessary, if users wanted to they can write their own helper functions

File: include/Util/Transform.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    /* wrapped operator */
+    Transform operator*(Transform const& other) { return Transform(m_mat3 * other.m_mat3); }
+    
+    /* unwrap (getter) */
+    glm::mat3 Mat3() const { return m_mat3; }
+protected:
+    glm::mat3 m_mat3;
```
Comment (NOOBDY) [inline include/Util/Transform.hpp:28]: I think calling this `m_Data` or something would be better

File: include/Util/Transform.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    Transform RotateByRadians(float radians) const;
+
+    Transform Translate(const glm::vec2& translation) const;
+    Transform Translate(float x, float y) const { return Translate(glm::vec2(x,y)); }
+
+    Transform Scale(const glm::vec2& scale) const;
+    Transform Scale(float x, float y) const { return Scale(glm::vec2(x,y)); }
+
+    /* wrapped operator */
+    Transform operator*(Transform const& other) { return Transform(m_mat3 * other.m_mat3); }
```
Comment (NOOBDY) [inline include/Util/Transform.hpp:23]: I still think this is a leaky abstraction

**Review by NOOBDY — CHANGES_REQUESTED** [review]

**Review by NOOBDY — CHANGES_REQUESTED** [review]

**Comment by NOOBDY** [comment]
we should add functions for resetting value, something like `SetTranslation(glm::vec2)`

**Review by NOOBDY — CHANGES_REQUESTED** [review]
- [ ] Add Setter
- [ ] Add Readable Getter
- [ ] `const Transform &other`
- [ ] Remove mat3 API

File: include/Util/Transform.hpp
Comment (NOOBDY) [inline include/Util/Transform.hpp:1]: Maybe change the order to TRS?

File: src/Util/Transform.cpp
Comment (NOOBDY) [inline src/Util/Transform.cpp:1]: Pro tip:
you can use
```
// NOLINTBEGIN(readability-identifier-length)

// code goes here

// NOLINTEND(readability-identifier-length)
```

File: src/Util/Transform.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    m_Mat3 = {
+        cos(er) * s.x,  sin(er) * s.y, 0,   //
+        -sin(er) * s.x, cos(er) * s.y, 0,   //
+        m_Mat3[2][0],   m_Mat3[2][1],  1.0f //
+    };                                      // dont forget transpose your view
+}
+
+float Transform::GetRotation() const {
+    auto radians = atan2(m_Mat3[0][1], m_Mat3[0][0]);
+    return radians >= 0 ? radians : radians + 4 * acos(0.0);
```
Comment (NOOBDY) [inline src/Util/Transform.cpp:56]: using `std::atan2` and `std::acos` would be more "C++" here

Comment (NOOBDY) [inline src/Util/Transform.cpp:56]: there are narrowing conversion warnings here

**Review by NOOBDY — CHANGES_REQUESTED** [review]

**Review by NOOBDY — CHANGES_REQUESTED** [review]
more things i noticed

File: src/Util/Transform.cpp
Comment (NOOBDY) [reply src/Util/Transform.cpp:1]: why didn't you use it here
`readability-identifier-length` and `readability-magic-numbers`

File: test/TransformTest.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    Transform transform = Transform().Translate(glm::vec2(6, 9));
+    EXPECT_EQ(transform.GetTranslation(), glm::vec2(6, 9));
+}
+
+TEST(TransformTest, Rotate360) {
+    auto expected = glm::radians(360.0F);
+    Transform transform = Transform().Rotate(expected);
+    auto result = transform.GetRotation();
+    // IDK the better way to fix v, is caused by PI approximation `\_(:/)_/`
+    EXPECT_NEAR(result + 4 * acos(0.0), expected, M_PI_TOLERANCE);
```
Comment (NOOBDY) [inline test/TransformTest.cpp:34]: missed `std::` here

**Review by NOOBDY — CHANGES_REQUESTED** [review]

File: test/TransformTest.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+TEST(TransformTest, SetGetRotation) {
+    auto expected = glm::radians(269.69F);
+    Transform transform = Transform();
+    transform.SetRotation(expected);
+    transform.SetScale(glm::vec2(6, 9));
+    auto result = transform.GetRotation();
+    // Note the 1000x tolerance
+    EXPECT_NEAR(result, expected, M_PI_TOLERANCE * 1e+3F);
```
[the remark this answers is not in the document]
Comment (ntut-rick) [reply test/TransformTest.cpp:69]: IDK, Maybe bc it have 1000x tolerance after 269.69 degrees rotation 🤷‍♀️

**Review by NOOBDY — DISMISSED** [review]
LGTM

File: src/Util/Transform.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    };                    // dont forget transpose your view
+    return {scaleMatrix * m_Mat3};
+}
+
+void Transform::SetRotation(float r) {
+    glm::vec2 es = GetScale(); // existing Scale
+    m_Mat3 = {
+        std::cos(r) * es.x,  std::sin(r) * es.y, 0,   //
+        -std::sin(r) * es.x, std::cos(r) * es.y, 0,   //
+        m_Mat3[2][0],   m_Mat3[2][1],  1.0f //
```
Comment (NOOBDY) [inline src/Util/Transform.cpp:40]:
```suggestion
        m_Mat3[2][0],   m_Mat3[2][1],  1.0F //
```

File: src/Util/Transform.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    m_Mat3[2][0] = t.x;
+    m_Mat3[2][1] = t.y;
+}
+
+void Transform::SetScale(const glm::vec2 &s) {
+    float er = GetRotation(); // existing Rotation
+    m_Mat3 = {
+        std::cos(er) * s.x,  std::sin(er) * s.y, 0,   //
+        -std::sin(er) * s.x, std::cos(er) * s.y, 0,   //
+        m_Mat3[2][0],   m_Mat3[2][1],  1.0f //
```
Comment (NOOBDY) [inline src/Util/Transform.cpp:54]:
```suggestion
        m_Mat3[2][0],   m_Mat3[2][1],  1.0F //
```

File: src/Util/Transform.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
+#include "Util/Transform.hpp"
+
+namespace Util {
+
+// NOLINTBEGIN(readability-identifier-length)
+// NOLINTBEGIN(readability-magic-numbers)
```
Comment (NOOBDY) [inline src/Util/Transform.cpp:6]:
```suggestion
// NOLINTBEGIN(readability-identifier-length,readability-magic-numbers)
```

File: src/Util/Transform.cpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    return {m_Mat3[2][0], m_Mat3[2][1]};
+}
+
+glm::vec2 Transform::GetScale() const {
+    return {glm::length(glm::vec2(m_Mat3[0][0], m_Mat3[1][0])),
+            glm::length(glm::vec2(m_Mat3[0][1], m_Mat3[1][1]))};
+}
+
+// NOLINTEND(readability-identifier-length)
+// NOLINTEND(readability-magic-numbers)
```
Comment (NOOBDY) [inline src/Util/Transform.cpp:73]:
```suggestion
// NOLINTEND(readability-identifier-length,readability-magic-numbers)
```

**Review by NOOBDY — CHANGES_REQUESTED** [review]

**Comment by Dragonfly911117** [comment]
The `add_compile_definitions(GLEW_NO_GLU)` option, which was adde in ea0ba3c,  in Dependencies.cmake is gone; and making it uncompile-able.

**Review by Dragonfly911117 — CHANGES_REQUESTED** [review]
everything else seems fine to me

**Review by NOOBDY — APPROVED** [review]

**Review by Dragonfly911117 — APPROVED** [review]
LGTM

**Pull request merged**